shipitscript/shipitscript/ship_actions.py

```diff
@@ -33,36 +33,83 @@ def get_shippable_revision(repo, last_shipped_revision):
+    # tail of the list is the most recent  version we have shipped based on
+    # https://github.com/mozilla-releng/shipit/blob/master/api/src/shipit_api/api.py#L131
+    try:
+        most_recent_release = all_releases[-1]
+        return most_recent_release['revision']
+    except IndexError:
+        # return None should the list is empty
+        log.error('The list of releases is empty')
+        return
+    except KeyError:
+        log.error('No `revision` key present in the most recent release')
+        return
+
+
+def calculate_build_number(shipit_config, product, branch, version):
     release_api, headers = get_shipit_api_instance(shipit_config)
 
+    log.info('Call Ship-it to retrieve all releases matching criteria ...')
+    all_releases = release_api.get_releases(
+        product,
+        branch,
+        status='shipped,aborted,scheduled',
+        version=version,
+        headers=headers,
+    )
+
+    build_numbers = [r['build_number'] for r in all_releases]
+    last_build_number = max(build_numbers) or int(0)
+    return last_build_number + 1
 
-def start_new_release(
-    self,
-    product,
-    channel,
-    release_name,
-    version,
-    revision,
-    phase,
-    shipit_config,
-    headers={},
-):
+
+def releases_are_disabled(shipit_config, product, branch):
     release_api, headers = get_shipit_api_instance(shipit_config)
-    release_name = ""
 
+    log.info('Call Ship-it to check for disabled products across branches')
+    disabled_products = release_api.get_disabled_products(headers=headers)
+
+    if product in disabled_products:
+        if branch in disabled_products[product]:
+            log.info(f'Product {product} and {branch} are currently disabled')
+            return True
+
+    log.info(f'Product {product} and {branch} is enabled. Continuing ...')
+    return False
+
+
+def start_new_release(shipit_config, product, branch, version, revision, phase):
+    # compute the build_number for the to-be-created release
+    build_number = calculate_build_number(shipit_config, product, branch, version)
+
+    # safeguard to avoid creating releases if they have been disabled from UI
+    if releases_are_disabled(shipit_config, product, branch):
+        return
+
+    release_api, headers = get_shipit_api_instance(shipit_config)
     log.info('creating a new release...')
-    release_api.create_new_release(
-        product, channel, release_name, version, revision, headers=headers
-    )
-    release_api.trigger_release_phase(
-        product, channel, release_name, phase, headers=headers
+    release_details = release_api.create_new_release(
+        product, branch, version, build_number, revision, headers=headers
     )
 
+    # grab the release name from the Ship-it response
+    release_name = release_details['name']
+
+    # avoid race conditions in between creating the release and triggering the
+    # specific `phase`
+    if releases_are_disabled(shipit_config, product, branch):
```

Comment:
You never know, maybe they will be disabled after you check here and before you trigger a phase? :P

Reply:
I've enriched the comment to say we're attempting to minimize the possiblity of a race condition. True that it can still occur. Is there a better way to solve this?